src/components/icon/icon.js

```diff
@@ -20,25 +21,221 @@ angular.module('material.components.icon', [
  * @restrict E
  *
  * @description
- * The `<md-icon>` directive is an element useful for SVG icons
+ * The `<md-icon>` directive is an element useful for showing an icon
  *
  * @usage
  * <hljs lang="html">
- *  <md-icon icon="/img/icons/ic_access_time_24px.svg">
- *  </md-icon>
+ *  <md-icon font-icon="android" alt="android icon"></md-icon>
+ *  <md-icon svg-icon="action:android" alt="android icon"></md-icon>
+ *  <md-icon svg-src="/android.svg" alt="android icon"></md-icon>
  * </hljs>
- *
  */
-function mdIconDirective() {
+function mdIconDirective($mdIcon, $mdAria, $log) {
   return {
+    scope: {
+      fontIcon: '@',
+      svgIcon: '@',
+      svgSrc: '@'
+    },
     restrict: 'E',
-    template: '<object class="md-icon"></object>',
-    compile: function(element, attr) {
-      var object = angular.element(element[0].children[0]);
-      if(angular.isDefined(attr.icon)) {
-        object.attr('data', attr.icon);
-      }
+    template: getTemplate,
+    link: postLink
+  };
+
+  function getTemplate(element, attr) {
+    return attr.fontIcon ? '<span class="md-font" ng-class="fontIcon"></span>' : '';
+  }
+
+  /**
+   * Directive postLink
+   * Supports embeded SVGs, font-icons, & external SVGs
+   */
+  function postLink(scope, element, attr) {
+    if (attr.alt == '') {
```

Comment:
We might want to use `scope.$watch` or `attr.$observe` here if we want to allow someone to have dynamic icons...

For example, I could see the following being a common use case:

```
<md-icon svg-icon="{{ selected ? 'checkbox-checked' : 'checkbox-unchecked' }}">
```

To help maintain performance we could register only one listener (depending on which mode they are using (`svgSrc`, `svgIcon` or `fontIcon`)